File: packages/dart/lib/src/location/rand_city.dart
import 'dart:math';

import 'package:randino/src/location/location_generator.dart';
import 'package:randino/src/types.dart';

/// Generate the divisions a region is made of: a Korean 시·군·구, written with its
/// city where it is one of a city's districts (`수원시 장안구`), or a US city,
/// town, village or census designated place.
///
/// A name can come back more than once from different regions — Korea has a
/// `중구` in five of them — and [unique] compares the names. A null [language]
/// means every language.
///
/// ```dart
/// randCity(language: LocationLanguage.ko, count: 3); // [강남구, 수원시 장안구, 양평군]
/// randCity(language: LocationLanguage.en, count: 2); // [Pasadena, Burlington]
/// ```
List<String> randCity({
  LocationLanguage? language,
  int count = 1,
  int? minLength,
  int? maxLength,
  String? startsWith,
  bool unique = false,

  /// Where the randomness comes from: `Random.secure()` for a value nobody may
  /// predict, `Random(42)` for one that has to come out the same every run.
  Random? random,
}) => [
  for (final detail in generateLocationDetails(
    form: LocationForm.unit,
    level: LocationLevel.city,
    language: language,
    count: count,
    minLength: minLength,
    maxLength: maxLength,
    startsWith: startsWith,
    unique: unique,
    random: random,
  ))
    detail.location,
];

File: packages/javascript/lib/location/randCity.ts
import type { LocationDetail, RandLocationUnitOptions } from '../_types/global.js';
import { drawLocation } from './locationGenerator.js';

/**
 * Generate the divisions a region is made of: a Korean 시·군·구, written with its
 * city where it is one of a city's districts (`수원시 장안구`), or a US city, town,
 * village or census designated place.
 *
 * A name can come back more than once from different regions — Korea has a
 * `중구` in five of them — and `unique` compares the names.
 *
 * @example
 * randCity({ language: 'ko', count: 3 }); // ['강남구', '수원시 장안구', '양평군']
 * randCity({ language: 'en', count: 2 }); // ['Pasadena', 'Burlington']
 */
export function randCity(options?: RandLocationUnitOptions & { output?: 'value' }): string[];
/**
 * Cities along with the region and country each one is in.
 *
 * @example
 * randCity({ language: 'ko', output: 'detail' });
 * // [{ location: '강남구', language: 'ko', level: 'city',
 * //    country: '대한민국', region: '서울특별시', city: '강남구', district: null }]
 */
export function randCity(options: RandLocationUnitOptions & { output: 'detail' }): LocationDetail[];
export function randCity(options: RandLocationUnitOptions = {}): string[] | LocationDetail[] {
	return drawLocation('unit', 'city', options);
}
